Rename variables es3 reserved words

maurobringolf opened the pull request:
<!-- 
Before making a PR please make sure to read our contributing guidelines 
https://github.com/babel/babel/blob/master/CONTRIBUTING.md

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. It should be underlined in the preview if done correctly.
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | Fixes #6477 
| Patch: Bug Fix?          | no
| Major: Breaking Change?  | no? 
| Minor: New Feature?      | yes
| Tests Added + Pass?      | Yes
| Documentation PR         | <!-- If so, add `[skip ci]` to your commit message to skip CI -->
| Any Dependency Changes?  | 

<!-- Describe your changes below in as much detail as possible -->

This is not done yet, but I would like to confirm that I understood the requirements of the issue correctly. The keyword lists included here are copy pasted directly from the specs.

Should I visit variable declarations instead of identifiers? **BindingIdentifier** and **ReferencedIdentifier** ✅ 

Also what about *Future reserved keywords* from the spec? Should these be optional or default?

xtuc commented:
That looks nice, good job. 

I think we could store the keywords in Babylon, which is already the case but there's not separation.

This reminds me of https://github.com/babel/babel/issues/5876. We could maybe use the plugin on our helpers before injecting them.

hzoo commented:
There should already be a dependency somewhere that has keywords check, I think `esutils` unless it's not sufficient

loganfsmyth commented on packages/babel-plugin-transform-es3-reserved-words/src/reserved-words.js:
```diff
@@ -0,0 +1,96 @@
+const es2017 = [
```
It would be great to add a test that checks each reserved word. We _may_ be able to use the same helper we do in the other `es3` transforms, but you'll have to verify. https://github.com/babel/babel/blob/5f285c1034acebf9303018a541831da143823cc4/packages/babel-plugin-transform-es3-property-literals/src/index.js#L10

maurobringolf replied:
As far as I can tell the helper from `babel-types` uses `esutils.keyword` and only checks for ES5 and ES6, but not ES3 keywords. That would mean that other ES3 plugins actually don't respect all ES3 keywords?

loganfsmyth replied:
Hmm, interesting. I think I'd be in favor of adding another utility to `babel-types` alongside the current one, but for ES3 identifiers. Then we can use that function across all three transforms and delete this file.

loganfsmyth commented on packages/babel-plugin-transform-es3-reserved-words/src/index.js:
```diff
@@ -0,0 +1,15 @@
+
+  return {
+    visitor: {
+      Identifier(path) {
```
`Identifier` also covers things like `.foo`, so this would probably get called more often than we want. I'd recommend having this visit the `BindingIdentifier` and `ReferencedIdentifier` types specifically.

Andarist commented on packages/babel-plugin-transform-es3-reserved-words/src/reserved-words.js:
```diff
@@ -0,0 +1,96 @@
+  "with",
+  "yield",
+];
+const es3 = [
```
could maybe `es2017` array be constructed by copying the `es3` + adding/removing neccessary items? such derived list is really helpful as documentation and its really easy to spot immediately what has changed between `es3` and `2017`

also I would sort `es3` array alphabetically (`es2017` is already sorted)

maurobringolf replied:
Alphabetically sorted `es3` ✅ 

Currently left it as two arrays because I'm still not sure if these lists are not already available. If they stay, I agree that we should have them build upon each other 👍

Andarist commented on packages/babel-plugin-transform-es3-reserved-words/src/index.js:
```diff
@@ -0,0 +1,15 @@
+import { es3, es2017 } from "./reserved-words";
+
+export default function() {
+  const reservedKeywords = es3.filter(w => !(es2017.indexOf(w) !== -1));
```
personally I would move this outside of the plugin function to the outer scope, if something is nested i'm always instantly searching for a reason why is it nested (basically what variables from the current closure are used) - and when it the things are top level is way more obvious that i can mentally just skip this step while reading

also what it does is a [`difference`](https://lodash.com/docs/4.17.4#difference) - im not saying we should add dependency on lodash (or any other util lib) here, but it would be beneficial to extract this function to give more meaning to whats going on here
```js
const difference = (arr, arr2) => arr.filter(el => !(arr2.indexOf(el) !== -1));
const reservedKeywords = difference(es3, es2017);
```

maurobringolf replied:
I really like this, thank you! Changed it.

loganfsmyth commented on packages/babel-plugin-transform-es3-reserved-words/src/index.js:
```diff
@@ -0,0 +1,16 @@
+  return {
+    visitor: {
+      "BindingIdentifier|ReferencedIdentifier"(path) {
+        if (reservedKeywords.indexOf(path.node.name) !== -1) {
```
Babel is targeting Node >= 4 now, so you can use `Set` with `.has` here instead of an array.

maurobringolf commented:
I tried implementing suggestions by @loganfsmyth and @Andarist: Moved ES3-only reserved words into a constant Set in `babel-types` and added a new helper `isValidES3Identifier`. In all three es3 plugins I swapped `isValidIdentifier` for `isValidES3Identifier`.

existentialism commented on packages/babel-plugin-transform-es3-reserved-words/package.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "babel-plugin-transform-es3-reserved-words",
+  "version": "7.0.0-beta.2",
```
Couple things:

1. Rename to `@babel/plugin-transform-es3-reserved-words`
2. Match version to `7.0.0-beta.3`
3. Set devDep to `@babel/helper-plugin-test-runner` and match its version to `7.0.0-beta.3`

nicolo-ribaudo commented on packages/babel-plugin-transform-es3-reserved-words/test/index.js:
```diff
@@ -0,0 +1,3 @@
+import runner from "babel-helper-plugin-test-runner";
```
Also here

maurobringolf replied:
Ahhh damn it, thanks!

maurobringolf replied:
Haha now CI fails the other way around 😀

nicolo-ribaudo replied:
Lol

nicolo-ribaudo commented:
@maurobringolf You need to rebase your branch on master, otherwise `@babel/helper-plugin-test-runner` can't be resolved (since it was renamed after you created your branch).

nicolo-ribaudo commented on packages/babel-plugin-transform-es3-reserved-words/test/fixtures/issue-6477/issue-6477/options.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "plugins": ["../../../../lib/index.js"]
```
After rebasing, can you try again to use `transform-es3-reserved-words` here?

maurobringolf commented:
Thank you @nicolo-ribaudo, looking good!

maurobringolf commented:
Im fine with it, for the question regarding different reserved word classes I think its perfectly fine to rename them all. My main concern was the word list inside the plugin, but I think this new helper in babel-types is a good solution. So yeah, ready from my perspective.

winston01 commented:
Hi

There is an issue in package.json: the `main` property is set to `lib/index.js`, while the correct location is `src/idex.js`. This causes `Error: Cannot find module '@babel/plugin-transform-es3-reserved-words'` while trying to use the module.

L
